### ISSUE TITLE
Refactor convergence trajectory tracking and harmonize loss-related APIs

### ISSUE THREAD

**Issue by jgallowa07** [matsengrp/multidms#195]
# Refactor convergence trajectory tracking and harmonize loss-related APIs

## Summary

Refactor how convergence metrics are tracked during fitting, and harmonize the loss-evaluation methods across `Model` and `ModelCollection`. The convergence trajectory should track three meaningful quantities (`objective_total`, `objective_error`, `loss`) directly from `jaxmodels.fit` as a DataFrame. The loss-evaluation methods (`conditional_loss`, `get_df_loss`) should be renamed and unified to return a consistent dict structure. The corresponding `ModelCollection` methods should be renamed to align.

### Convergence trajectory problems

- `jaxmodels.fit` computes three distinct quantities at each block coordinate descent iteration:
  1. **`objective_total`**: the full objective value (loss + L2 + fusion + beta0_ridge penalties), currently computed with a scaling factor (`objective_total(..., scale=scale)` at line 701).
  2. **`objective_error`**: the relative change in the objective between iterations (`abs(obj_old - obj) / max(abs(obj_old), abs(obj), 1)` at line 711).
  3. **`loss`**: the raw loss function value (`sum(loss_fn(model, data_sets, **loss_kwargs).values())`), currently only computed *inside* the objective functions but never tracked independently.

- Currently, only the **scaled** `objective_total` value is appended to `loss_trajectory` (line 716) and returned. `Model.convergence_trajectory_df` then misleadingly names this column `"loss"` and derives an `"error"` column as `diff().abs()` — which is neither the actual loss nor the actual convergence error.

### Loss evaluation API problems

- `Model.conditional_loss` (property): evaluates `loss_fn` on **training data**, returns `{condition: loss}` dict — missing the total.
- `Model.get_df_loss(df, conditional=False)`: evaluates `loss_fn` on **arbitrary data**, returns either a `{condition: loss}` dict or a scalar depending on the `conditional` flag — inconsistent return types and confusing name.
- `ModelCollection.__init__` stores per-condition training losses as `{condition}_loss_training` columns but not the total.
- `ModelCollection.add_validation_loss` stores both per-condition and total validation losses, creating asymmetry with training.
- `ModelCollection.get_conditional_loss_df` melts these columns into long form — the name is verbose and doesn't convey what it returns.

### Part 1: Convergence trajectory

1. In `jaxmodels.fit`, build a `pd.DataFrame` with columns `iteration`, `objective_total_trajectory`, `objective_error_trajectory`, and `loss_trajectory`. Track all three quantities at each block iteration. Store the **unscaled** `objective_total` (call with `scale=1.0`).
2. Return this DataFrame from `jaxmodels.fit` instead of the current `list[float]`.
3. In `Model`, store this DataFrame directly and return it from `convergence_trajectory_df`.
4. Update `Model.converged` to use the `objective_error_trajectory` column from the trajectory DataFrame.

### Part 2: Loss evaluation API

1. Rename `Model.conditional_loss` → `Model.training_loss` (property). Returns `{condition_1: loss, condition_2: loss, ..., "total": loss}`.
2. Rename `Model.get_df_loss` → `Model.eval_loss(df)` (method). Always returns `{condition_1: loss, condition_2: loss, ..., "total": loss}`. Remove the `conditional` parameter.
3. Place these two methods adjacent in the source file.

### Part 3: ModelCollection harmonization

1. Rename `ModelCollection.add_validation_loss` → `ModelCollection.add_eval_loss`. Internally calls `Model.eval_loss`.
2. Rename `ModelCollection.get_conditional_loss_df` → `ModelCollection.loss_df`. Returns the same long-form DataFrame.
3. In `ModelCollection.__init__`, also store `total_loss_training` (matching the validation pattern).
4. `ModelCollection.convergence_trajectory_df` — name stays, calls `Model.convergence_trajectory_df` internally.

**Key design decisions:**
- All three trajectory values are tracked because they serve distinct purposes: `objective_total` for absolute objective magnitude, `objective_error` for convergence monitoring, `loss` for the raw data-fit quality without regularization.
- Unscaled `objective_total` is stored because the scale factor is an internal numerical convenience, not meaningful to the user.
- Consistent dict return type `{condition: loss, "total": loss}` from both `training_loss` and `eval_loss` eliminates the confusing `conditional` flag and makes the API uniform.
- `"_trajectory"` suffix on column names in the convergence DataFrame distinguishes these time-series quantities from point-in-time values.

### Programmatic API — Convergence trajectory

```python
# Before (current)
model.convergence_trajectory_df
#   iteration    loss     error
# 0         0  1.2345  0.000000
# 1         1  0.9876  0.246900

# After (proposed)
model.convergence_trajectory_df
#   iteration  objective_total_trajectory  objective_error_trajectory  loss_trajectory
# 0         0                    1234.500                   0.000000          980.300
# 1         1                     987.600                   0.002340          776.100
```

### Programmatic API — Loss evaluation

```python
# Before
model.conditional_loss  # -> {"cond_a": 1.23, "cond_b": 4.56}
model.get_df_loss(df)   # -> 5.79 (float)
model.get_df_loss(df, conditional=True)  # -> {"cond_a": 1.23, "cond_b": 4.56}

# After
model.training_loss  # -> {"cond_a": 1.23, "cond_b": 4.56, "total": 5.79}
model.eval_loss(df)  # -> {"cond_a": 1.23, "cond_b": 4.56, "total": 5.79}
```

### Programmatic API — ModelCollection

```python
# Before
mc.add_validation_loss(test_df)
mc.get_conditional_loss_df()

# After
mc.add_eval_loss(test_df)
mc.loss_df()
```

### 1. `jaxmodels.fit` — return convergence trajectory DataFrame

**File:** `multidms/jaxmodels.py` (existing)

Build a DataFrame tracking all three quantities at each iteration. The `objective_total` should be **unscaled** (pass `scale=1.0`). The `loss` is computed by calling `loss_fn` directly.

```python
import pandas as pd

# Initialize trajectory storage
trajectory_rows = []

# Inside the fitting loop, after the existing obj and objective_error computations:
obj_unscaled = objective_total(
    model, data_sets, l2reg=l2reg, fusionreg=fusionreg,
    scale=1.0, beta0_ridge=beta0_ridge
)
loss = sum(loss_fn(model, data_sets, **loss_kwargs).values())

trajectory_rows.append({
    "iteration": k,
    "objective_total_trajectory": float(obj_unscaled),
    "objective_error_trajectory": float(objective_error),
    "loss_trajectory": float(loss),
})

# At end, return DataFrame
convergence_trajectory_df = pd.DataFrame(trajectory_rows)
return model, convergence_trajectory_df
```

Update the return type annotation from `tuple[Model, list[float]]` to `tuple[Model, pd.DataFrame]`.

### 2. `Model.fit` — store convergence trajectory DataFrame

**File:** `multidms/model.py` (existing)

```python
# In __init__:
self._convergence_trajectory_df = None  # replaces self._loss_trajectory

# In fit():
self._jax_model, self._convergence_trajectory_df = jaxmodels.fit(...)
```

### 3. `Model.convergence_trajectory_df` — return stored DataFrame directly

**File:** `multidms/model.py` (existing)

```python
@property
def convergence_trajectory_df(self) -> pd.DataFrame:
    """Convergence trajectory from fitting.

    Returns
    -------
    pd.DataFrame or None
        DataFrame with columns 'iteration', 'objective_total_trajectory',
        'objective_error_trajectory', 'loss_trajectory'.
        None if model has not been fitted.
    """
    return self._convergence_trajectory_df
```

### 4. `Model.converged` — use trajectory DataFrame

**File:** `multidms/model.py` (existing)

```python
@property
def converged(self) -> bool:
    """Whether the model fitting converged."""
    if self._convergence_trajectory_df is None or len(self._convergence_trajectory_df) < 2:
        return False
    last_error = self._convergence_trajectory_df["objective_error_trajectory"].iloc[-1]
    return last_error < self._fit_tol
```

### 5. `Model.training_loss` — rename from `conditional_loss`

**File:** `multidms/model.py` (existing)

```python
@property
def training_loss(self) -> dict:
    """Per-condition and total loss on training data.

    Returns
    -------
    dict[str, float]
        Dictionary mapping condition names to their training loss values,
        plus a "total" key with the sum.
    """
    if self._jax_model is None:
        raise ValueError("Model has not been fitted. Call fit() first.")
    loss_dict = self._loss_fn(
        self._jax_model, self._jax_data_sets, **self._loss_kwargs
    )
    result = {k: float(v) for k, v in loss_dict.items()}
    result["total"] = sum(result.values())
    return result
```

### 6. `Model.eval_loss` — rename from `get_df_loss`

**File:** `multidms/model.py` (existing)

Remove the `conditional` parameter. Always return `{condition: loss, "total": loss}`.

```python
def eval_loss(self, df) -> dict:
    """Evaluate the model's loss on an arbitrary DataFrame.

    Parameters
    ----------
    df : pd.DataFrame
        DataFrame with columns 'condition', 'aa_substitutions', 'func_score'.

    Returns
    -------
    dict[str, float]
        Per-condition losses plus a "total" key with the sum.
    """
    # ... (existing validation logic unchanged) ...
    encoded = self._encode_variants(df)
    temp_data_sets = {condition: data for condition, (data, _) in encoded.items()}
    loss_dict = self._loss_fn(self._jax_model, temp_data_sets, **self._loss_kwargs)
    result = {k: float(v) for k, v in loss_dict.items()}
    result["total"] = sum(result.values())
    return result
```

Place `training_loss` and `eval_loss` adjacent in the source file.

### 7. `ModelCollection.__init__` — store total training loss

**File:** `multidms/model_collection.py` (existing)

```python
# In __init__, after computing per-condition training losses:
for idx, fit in fit_models.iterrows():
    training_loss = fit.model.training_loss
    for condition, loss in training_loss.items():
        if condition == "total":
            fit_models.loc[idx, "total_loss_training"] = loss
        else:
            fit_models.loc[idx, f"{condition}_loss_training"] = loss
```

### 8. `ModelCollection.add_eval_loss` — rename from `add_validation_loss`

**File:** `multidms/model_collection.py` (existing)

Rename method, update internal call from `get_df_loss(df, conditional=True)` to `eval_loss(df)`.

```python
def add_eval_loss(self, test_data, overwrite=False):
    """Add evaluation (validation) loss to the fit collection dataframe."""
    # ... existing logic ...
    for idx, fit in self.fit_models.iterrows():
        eval_result = fit.model.eval_loss(test_data[fit["dataset_name"]])
        for condition, loss in eval_result.items():
            if condition == "total":
                self.fit_models.loc[idx, "total_loss_validation"] = loss
            else:
                self.fit_models.loc[idx, f"{condition}_loss_validation"] = loss
```

### 9. `ModelCollection.loss_df` — rename from `get_conditional_loss_df`

**File:** `multidms/model_collection.py` (existing)

Rename method. Logic stays the same (melts the `*_loss_training` and `*_loss_validation` columns from `fit_models`).

```python
def loss_df(self, query=None):
    """Return a long-form DataFrame of training and validation losses.

    Columns: 'dataset_name', 'fusionreg', 'condition', 'loss', 'split'.
    """
    # ... same melt logic as current get_conditional_loss_df ...
```

### Data flow

```
jaxmodels.fit()
  └── returns (Model, pd.DataFrame)  # trajectory DF with 3 columns + iteration
        │
        ▼
Model.fit()
  └── stores self._convergence_trajectory_df
        │
        ├── Model.convergence_trajectory_df  # property, returns the DF
        ├── Model.converged                  # reads objective_error_trajectory
        │
        ▼
ModelCollection.convergence_trajectory_df()
  └── concatenates Model.convergence_trajectory_df across fits
```

```
Model.training_loss  # evaluates loss_fn on training data → {cond: val, "total": val}
Model.eval_loss(df)  # evaluates loss_fn on arbitrary df → {cond: val, "total": val}
        │
        ▼
ModelCollection.__init__         # calls training_loss → stores {cond}_loss_training + total
ModelCollection.add_eval_loss()  # calls eval_loss → stores {cond}_loss_validation + total
ModelCollection.loss_df()        # melts stored columns into long-form DF
```

### Note on `objective_error` at iteration 0

At iteration 0, there is no `obj_old` for computing `objective_error`. Looking at the current code, `obj_old` is computed at the *start* of each iteration (line 578), then `objective_error` is computed at the *end* (line 711) after all blocks run. So at iteration 0, `objective_error` measures the change from the model's initial state to its state after the first round of block coordinate descent. This is well-defined and should be stored as-is.

## Testing Strategy

- Update `test_model_fit_convergence_trajectory` to assert columns `iteration`, `objective_total_trajectory`, `objective_error_trajectory`, `loss_trajectory`
- Test `Model.training_loss` returns dict with per-condition keys plus `"total"`
- Test `Model.eval_loss(df)` returns dict with per-condition keys plus `"total"`
- Test that `Model.training_loss` values match `Model.eval_loss(training_df)` for the same data
- Test `ModelCollection.__init__` stores `total_loss_training`
- Test `ModelCollection.add_eval_loss` stores per-condition + total validation losses
- Test `ModelCollection.loss_df` returns expected columns
- Test `ModelCollection.convergence_trajectory_df` includes the new columns
- Update existing `ModelCollection` tests that reference old method names
- Edge cases:
  - Model not yet fitted (returns None / raises ValueError as appropriate)
  - Single iteration trajectory

## Documentation Updates

- [ ] Docstrings for all renamed/modified methods
- [ ] Notebooks left as-is (out of scope)

## Dependencies

None — no new dependencies required.

### Alternative 1: Keep trajectory as list, build DataFrame in Model

Build the DataFrame in `Model.convergence_trajectory_df` rather than in `jaxmodels.fit`.

**Why not chosen:** Building the DataFrame in `jaxmodels.fit` is cleaner because all three quantities are naturally available there. Passing three separate lists back to `Model` and assembling a DataFrame is more error-prone.

### Alternative 2: Keep `conditional` parameter on `eval_loss`

Maintain backward compatibility with the `conditional=False` default that returns a scalar.

**Why not chosen:** Having two return types from the same method is confusing. Always returning a dict with `"total"` is simpler and the caller can easily do `result["total"]` if they want just the sum.

### Alternative 3: Keep existing method names

Don't rename `conditional_loss`, `get_df_loss`, etc.

**Why not chosen:** The current names don't clearly convey what data the loss is computed on (training vs arbitrary) or what the return type is. `training_loss` / `eval_loss` are more descriptive and form a natural pair.

## Success Criteria

- [ ] `jaxmodels.fit` returns `(Model, pd.DataFrame)` with columns `iteration`, `objective_total_trajectory`, `objective_error_trajectory`, `loss_trajectory`
- [ ] `objective_total_trajectory` contains **unscaled** objective values
- [ ] `Model.convergence_trajectory_df` returns the DataFrame directly
- [ ] `Model.converged` uses `objective_error_trajectory` from the DataFrame
- [ ] `Model.training_loss` returns `{condition: loss, "total": loss}` for training data
- [ ] `Model.eval_loss(df)` returns `{condition: loss, "total": loss}` for arbitrary data
- [ ] `ModelCollection.__init__` stores `total_loss_training` alongside per-condition training losses
- [ ] `ModelCollection.add_eval_loss` replaces `add_validation_loss`
- [ ] `ModelCollection.loss_df` replaces `get_conditional_loss_df`
- [ ] All tests updated and passing

## Future Work

- Update example notebooks to use new column names and method names
- Consider deprecation warnings for old method names during a transition period
